engine/src/main/java/org/hibernate/search/v6poc/logging/impl/RootFailureCollector.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.v6poc.logging.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.StringJoiner;
+
+import org.hibernate.search.v6poc.logging.spi.ContextualFailureCollector;
+import org.hibernate.search.v6poc.logging.spi.FailureCollector;
+import org.hibernate.search.v6poc.logging.spi.FailureContextElement;
+import org.hibernate.search.v6poc.util.impl.common.LoggerFactory;
+import org.hibernate.search.v6poc.util.impl.common.ToStringStyle;
+import org.hibernate.search.v6poc.util.impl.common.ToStringTreeBuilder;
+
+import org.jboss.logging.Messages;
+
+public class RootFailureCollector implements FailureCollector {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private static final FailureContextMessages MESSAGES = Messages.getBundle( FailureContextMessages.class );
+
+	/**
+	 * This prevents Hibernate Search from trying too hard to collect errors,
+	 * which could be a problem when there is something fundamentally wrong
+	 * that will cause almost every operation to fail.
+	 */
+	private final int failureLimit;
+
+	private FailureCollectorImpl delegate;
+	private int failureCount = 0;
+
+	public RootFailureCollector(int failureLimit) {
+		this.failureLimit = failureLimit;
+	}
+
+	public void checkNoFailure() {
+		if ( failureCount > 0 ) {
+			String renderedFailures = renderFailures();
+			throw log.bootstrapCollectedFailures( renderedFailures );
+		}
+	}
+
+	private String renderFailures() {
+		ToStringStyle style = ToStringStyle.multilineIndentStructure(
+				MESSAGES.contextFailuresSeparator(),
+				MESSAGES.contextIndent(),
+				MESSAGES.contextFailuresBulletPoint(),
+				MESSAGES.contextFailuresNoBulletPoint()
```

Comment:
I didn't know if languages such as Japanese, Corean, Arabic, Chinese use the same symbols as we do for bullet points and colons. Also didn't know whether the chosen separators would work well with right-to-left languages. So I just took the safest path.
It will only ever happen on failure anyway, and it's just a few strings. Not a big deal.